### DISCUSSION TITLE
Why do Python floating points come out with many extra digits (but not string representations)?

### DISCUSSION THREAD

**Discussion by lamhktommy** [sqlalchemy/sqlalchemy#9247] [Usage Questions] [answered]
### Describe the bug

I make use of 2 different ways to insert floating point data into my postgresql table, but they got different behaviours. 

Method 1 (which results floating point error shown in psql shell): 
```
async def insert():
    async with session:
        session.add(Data(id=1, data_value=100.1231231313))
        await session.commit()
```
Result:
![image](https://user-images.githubusercontent.com/35717697/216977564-8fee40ea-f21a-40c3-8885-d7e17f7fca71.png)

Method 2 (which i can view the exact value):
```
async def insert_raw():
    async with session:
        await session.execute(text("insert into data_table values(2, 100.1231231313)"))
        await session.commit()
```
Result:
![image](https://user-images.githubusercontent.com/35717697/216977912-1af14101-d8b5-46b1-8269-37b13303adf8.png)

What I expect is that I use method 1 will return as same as method 2 in DB

I put my code here: https://github.com/lamhktommy/test_sqlalchemy

### To Reproduce

```python
Python version: 3.9 
Postgresql version: 14 
Code: https://github.com/lamhktommy/test_sqlalchemy
1. pip install -r requirements.txt
2. create .env with your DB credentials (see .env.example)
3. alembic upgrade head (create tables)
4. run main.py (switch between insert_raw() and insert())
```

### Error

```
No Error Msg
``` 

### Versions

- OS: Windows 10
- Python: 3.9
- SQLAlchemy: 1.4
- Database: postgresql 14
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): asyncpg

**Comment by CaselIT** [comment]
Hi,

It's not clear what is the issue here

**Comment by lamhktommy** [comment]
Hi, I have updated above comment. Hope this can add more info for you

**Comment by zzzeek** [comment, marked as answer]
hi there -

Floating point numbers are not exact and use an approximation scheme in order to represent a value.  For Python specifically, the behavior you are seeing is described at https://docs.python.org/3/tutorial/floatingpoint.html .

in this case, your test is sending two very different numbers to the database; one is an exact string representation, the other is a lossy Python floating point.   


Here's an asyncpg program that can illustrate the difference.    

```py
import asyncio
from decimal import Decimal

import asyncpg



async def main():
    conn = await asyncpg.connect(dsn="postgres://scott:tiger@localhost/test")

    await conn.execute("drop table data_table")
    await conn.execute("create table if not exists data_table(id serial, data numeric)")

    stmt = await conn.prepare("insert into data_table (data) values($1)")
    await stmt.executemany([
        (100.1231231313, ),
        (Decimal("100.1231231313"), )
    ])


asyncio.run(main())
```

psql output:

```sql
test=# select * from data_table;
 id |                        data                        
----+----------------------------------------------------
  1 | 100.1231231313000051841299864463508129119873046875
  2 |                                     100.1231231313
(2 rows)

```

I hope this clarifies!

**Reply by lamhktommy** [reply]
so i must convert to Decimal object first if I want it become 100.1231231313?

**Reply by zzzeek** [reply]
i wouldnt convert it, I would not use python floating point representation in the first place

`Decimal("100.1231231313")` and `Decimal(100.1231231313)` are not the same thing (try it out)

**Reply by lamhktommy** [reply]
so in sqlalchemy, what ways can i do if I want 100.1231231313 in db using ORM instead of raw SQL?

**Reply by CaselIT** [reply]
where are you getting the value from? are you reading it from a file?

**Reply by lamhktommy** [reply]
through data feed with 4-8 decimal places for different items

**Reply by zzzeek** [reply]
> so in sqlalchemy, what ways can i do if I want 100.1231231313 in db using ORM instead of raw SQL?

use `Decimal` with a string in the same way

    obj = MyObject(data=Decimal("149.43545"))

**Reply by CaselIT** [reply]
> through data feed with 4-8 decimal places for different items

like mike said use decimal, without first converting it to flat
